### PR TITLE
Stream geometry information

### LINKED ISSUE #34: Support streaming NeXus geometry
If available in stream Mantid would make use of this information instead of loading an IDF for the instrument.

This is part of the work required to support streaming geometry for Torben. Also need to convert IDF from McStas to NeXus geometry, or directly implement NeXus geometry (+ `Nxevent_data`) in McStas?

### PR DESCRIPTION
**Pull request by matthew-d-jones**
### Description of work

NeXus-Streamer can now optionally include the `nexus_structure` json description of the NeXus file in the run start message. This allows streaming geometry information to Mantid.
See updates to READMEs for details.

A new schema is used the the run messages, as it has not been accepted into `master` of `streaming-data-types` yet I have created a `testing` conan schema package for now: `streaming-data-types/70e3dd1@ess-dmsc/testing`

### Issue

Closes #34

### Acceptance Criteria

Running via `docker-compose up` should demonstrate the new functionality, but you'll need to manually copy https://github.com/ess-dmsc/streaming-data-types/pull/41/files#diff-bc2e07b65de26ece673bcb48fc0cd482 into kafkacow's schemas directory to check the run start message contains the new nexus_structure field.

### Unit Tests

Updated run message unit tests and added new tests in nexus_prodcuer/test/JSONDescriptionLoaderTest.cpp

---

## Code Review (To be filled in by the reviewer only)

- [ ] Is the code of an acceptable quality?
- [ ] Do the changes function as described and is it robust?

---

## Nominate for Group Code Review (Anyone can nominate it)
Indicate if you think the code should be reviewed in a Thursday code review session.

- [ ] Recommend for group code review

Also, nominate it on the code_review Slack channel (does someone want to automate this?).

### CODE REVIEW

File: nexus_producer/src/NexusPublisher.cpp [hunk 1]
```diff
@@ -287,7 +291,7 @@ size_t NexusPublisher::createAndSendRunMessage(int runNumber) {
  * @param rawbuf - a buffer for the message
  * @return - size of the buffer
  */
-size_t NexusPublisher::createAndSendRunStopMessage() {
+size_t NexusPublisher::createAndSendRunStopMessage(int runNumber) {
```
Comment (rerpha) [inline nexus_producer/src/NexusPublisher.cpp]: i can;t remember if const-ref for ints is trivial? but if not may be worth adding it here.

File: nexus_producer/src/NexusPublisher.cpp [hunk 2]
```diff
@@ -101,7 +100,11 @@ RunData NexusPublisher::createRunMessageData(int runNumber) {
   auto runData = RunData();
   runData.setNumberOfPeriods(m_fileReader->getNumberOfPeriods());
   runData.setInstrumentName(m_fileReader->getInstrumentName());
-  runData.setRunNumber(runNumber);
+  runData.setRunID(std::to_string(runNumber));
+  if (!m_settings.jsonDescription.empty()) {
+    runData.setNexusStructure(JSONDescriptionLoader::loadJsonDescription(
```
Comment (rerpha) [inline nexus_producer/src/NexusPublisher.cpp]: Can you add a test that covers this? I know it will slightly duplicate the other tests, but it would be good to make sure that the rundata is as expected still even with JSON data in.

File: nexus_producer/src/JSONDescriptionLoader.cpp [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    position += replace.length();
+  }
+}
+
+std::string loadFromFile(const std::string &filepath) {
+  std::ifstream filestream(filepath);
+  std::string json_string;
+
+  filestream.seekg(0, std::ios::end);
+  json_string.reserve(filestream.tellg());
```
Comment (rerpha) [inline nexus_producer/src/JSONDescriptionLoader.cpp]: clang-tidy is telling me that we are passing a long from `filestream.tellg()` and we should be using unsigned long, casting it solves this warning: 
`json_string.reserve(static_cast<unsigned long>(filestream.tellg()));`

File: nexus_producer/src/NexusPublisher.cpp (hunk 1 above)
Comment (matthew-d-jones) [reply nexus_producer/src/NexusPublisher.cpp]: Not worth ref for ints but I should have been using `const` here.

File: nexus_producer/src/JSONDescriptionLoader.cpp [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+  checkIsValidJson(json_string);
+  return json_string;
+}
+
+std::string loadJsonDescription(const std::string &filepath,
+                                const std::string &instrumentName) {
+  std::string description = loadFromFile(filepath);
+  auto topicNames = TopicNames(instrumentName);
+  replaceString(description, "SAMPLE_ENV_TOPIC", topicNames.sampleEnv);
```
[the remark this answers is not in the document]
Comment (matthew-d-jones) [reply nexus_producer/src/JSONDescriptionLoader.cpp]: :+1:

File: nexus_producer/src/JSONDescriptionLoader.cpp (hunk 3 above)
Comment (matthew-d-jones) [reply nexus_producer/src/JSONDescriptionLoader.cpp]: :+1:

**Comment by rerpha** [comment]
Codecov was being stupid with it's patch comments so looks like CI has failed but i hasn't actually.
